File: ios/Processors/FaceProcessor.swift
//
//  LivenessCheckProcessor.swift
//  AzifaceMobileSdk
//
//  Created by Nayara Dias 08/03/2024.
//  Copyright © 2024 Azify. All rights reserved.
//

import UIKit
import Foundation
import FaceTecSDK

class FaceProcessor: NSObject, Processor, FaceTecFaceScanProcessorDelegate, URLSessionTaskDelegate {
    private let key: String!
    private let faceConfig: FaceConfig!
    private let CapThemeUtils: ThemeUtils! = ThemeUtils();
    var success = false
    var latestNetworkRequest: URLSessionTask!
    var fromViewController: AziFaceViewController!
    var faceScanResultCallback: FaceTecFaceScanResultCallback!

    init(sessionToken: String, fromViewController: AziFaceViewController, faceConfig: FaceConfig) {
        self.fromViewController = fromViewController
        self.faceConfig = faceConfig
        self.key = faceConfig.getKey()
        super.init()

        AzifaceMobileSdk.emitter.sendEvent(withName: "onCloseModal", body: true);

        let faceViewController = FaceTec.sdk.createSessionVC(faceScanProcessorDelegate: self, sessionToken: sessionToken)

        FaceTecUtilities.getTopMostViewController()?.present(faceViewController, animated: true, completion: nil)
    }

    func processSessionWhileFaceTecSDKWaits(sessionResult: FaceTecSessionResult, faceScanResultCallback: FaceTecFaceScanResultCallback) {
        fromViewController.setLatestSessionResult(sessionResult: sessionResult)

        self.faceScanResultCallback = faceScanResultCallback

        if sessionResult.status != FaceTecSessionStatus.sessionCompletedSuccessfully {
            if latestNetworkRequest != nil {
                latestNetworkRequest.cancel()
            }

            AzifaceMobileSdk.emitter.sendEvent(withName: "onCloseModal", body: false);
            faceScanResultCallback.onFaceScanResultCancel()
            return
        }

        var parameters: [String : Any] = [:]
        let extraParameters: [String: Any]? = self.faceConfig.getParameters()
        if extraParameters != nil {
            parameters["data"] = extraParameters
        }
        parameters["faceScan"] = sessionResult.faceScanBase64
        parameters["auditTrailImage"] = sessionResult.auditTrailCompressedBase64![0]
        parameters["lowQualityAuditTrailImage"] = sessionResult.lowQualityAuditTrailCompressedBase64![0]
        
        let hasExternalDatabaseRefID: Bool = self.faceConfig.getHasExternalDatabaseRefID()
        if hasExternalDatabaseRefID {
            parameters["externalDatabaseRefID"] = fromViewController.getLatestExternalDatabaseRefID()
        }

        let endpoint: String? = self.faceConfig.getEndpoint()
        var request = Config.makeRequest(url: endpoint ?? "", httpMethod: "POST")
        request.httpBody = try! JSONSerialization.data(withJSONObject: parameters, options: JSONSerialization.WritingOptions(rawValue: 0))

        let session = URLSession(configuration: URLSessionConfiguration.default, delegate: self, delegateQueue: OperationQueue.main)
        latestNetworkRequest = session.dataTask(with: request as URLRequest, completionHandler: { data, response, error in
            if let httpResponse = response as? HTTPURLResponse {
                if httpResponse.statusCode < 200 || httpResponse.statusCode >= 299 {
                    AzifaceMobileSdk.emitter.sendEvent(withName: "onCloseModal", body: false);
                    faceScanResultCallback.onFaceScanResultCancel()
                    return
                }
            }

            if let error = error {
                AzifaceMobileSdk.emitter.sendEvent(withName: "onCloseModal", body: false);
                faceScanResultCallback.onFaceScanResultCancel()
                return
            }

            guard let data = data else {
                AzifaceMobileSdk.emitter.sendEvent(withName: "onCloseModal", body: false);
                faceScanResultCallback.onFaceScanResultCancel()
                return
            }

            guard let responseJSON = try? JSONSerialization.jsonObject(with: data, options: JSONSerialization.ReadingOptions.allowFragments) as? [String: AnyObject] else {
                AzifaceMobileSdk.emitter.sendEvent(withName: "onCloseModal", body: false);
                faceScanResultCallback.onFaceScanResultCancel()
                return
            }

            guard let scanResultBlob = responseJSON["scanResultBlob"] as? String,
                  let wasProcessed = responseJSON["wasProcessed"] as? Bool else {
                AzifaceMobileSdk.emitter.sendEvent(withName: "onCloseModal", body: false);
                faceScanResultCallback.onFaceScanResultCancel()
                return;
            }

            if wasProcessed == true {
                let successMessage: String? = self.faceConfig.getSuccessMessage()
                let message = self.CapThemeUtils.handleMessage(self.key, child: "successMessage", defaultMessage: successMessage ?? "");
                FaceTecCustomization.setOverrideResultScreenSuccessMessage(message);

                self.success = faceScanResultCallback.onFaceScanGoToNextStep(scanResultBlob: scanResultBlob);
            } else {
                AzifaceMobileSdk.emitter.sendEvent(withName: "onCloseModal", body: false);
                faceScanResultCallback.onFaceScanResultCancel()
                return;
            }
        })

        latestNetworkRequest.resume()

        DispatchQueue.main.asyncAfter(deadline: .now() + 6) {
            if self.latestNetworkRequest.state == .completed { return }

            let uploadMessage: String = self.faceConfig.getUploadMessage()
            let message = self.CapThemeUtils.handleMessage(self.key, child: "uploadMessageIos", defaultMessage: uploadMessage);
            let uploadMessageOverride: NSMutableAttributedString = NSMutableAttributedString.init(string: message);
            faceScanResultCallback.onFaceScanUploadMessageOverride(uploadMessageOverride: uploadMessageOverride);
        }
    }

    func urlSession(_ session: URLSession, task: URLSessionTask, didSendBodyData bytesSent: Int64, totalBytesSent: Int64, totalBytesExpectedToSend: Int64) {
        let uploadProgress: Float = Float(totalBytesSent) / Float(totalBytesExpectedToSend)
        faceScanResultCallback.onFaceScanUploadProgress(uploadedPercent: uploadProgress)
    }

    func onFaceTecSDKCompletelyDone() {
        self.fromViewController.onComplete();
    }

    func isSuccess() -> Bool {
        return success
    }
}

File: android/src/main/java/com/azify/processors/FaceProcessor.java
package com.azify.processors;

import android.content.Context;

import androidx.annotation.NonNull;

import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.MediaType;
import okhttp3.RequestBody;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.util.Map;

import com.azify.processors.helpers.ThemeUtils;
import com.azify.azifacemobilesdk.AzifaceMobileSdkModule;
import com.facetec.sdk.*;

public class FaceProcessor extends Processor implements FaceTecFaceScanProcessor {
  private final String key;
  private final AzifaceMobileSdkModule aziFaceModule;
  private final FaceConfig faceConfig;
  private final ThemeUtils aziThemeUtils = new ThemeUtils();
  private boolean success = false;

  public FaceProcessor(String sessionToken, Context context, AzifaceMobileSdkModule aziFaceModule,
      FaceConfig faceConfig) {
    this.aziFaceModule = aziFaceModule;
    this.faceConfig = faceConfig;
    this.key = faceConfig.getKey();

    aziFaceModule.sendEvent("onCloseModal", true);
    FaceTecSessionActivity.createAndLaunchSession(context, FaceProcessor.this, sessionToken);
  }

  public void processSessionWhileFaceTecSDKWaits(final FaceTecSessionResult sessionResult,
      final FaceTecFaceScanResultCallback faceScanResultCallback) {
    aziFaceModule.setLatestSessionResult(sessionResult);

    if (sessionResult.getStatus() != FaceTecSessionStatus.SESSION_COMPLETED_SUCCESSFULLY) {
      NetworkingHelpers.cancelPendingRequests();
      faceScanResultCallback.cancel();
      aziFaceModule.sendEvent("onCloseModal", false);
      aziFaceModule.processorPromise.reject("The session status has not been completed!", "AziFaceInvalidSession");
      return;
    }

    JSONObject parameters = new JSONObject();
    try {
      final Map extraParameters = this.faceConfig.getParameters();
      if (extraParameters != null) {
        parameters.put("data", new JSONObject(extraParameters));
      }
      parameters.put("faceScan", sessionResult.getFaceScanBase64());
      parameters.put("auditTrailImage", sessionResult.getAuditTrailCompressedBase64()[0]);
      parameters.put("lowQualityAuditTrailImage", sessionResult.getLowQualityAuditTrailCompressedBase64()[0]);

      final boolean hasExternalDatabaseRefID = this.faceConfig.getHasExternalDatabaseRefID();
      if (hasExternalDatabaseRefID) {
        parameters.put("externalDatabaseRefID", aziFaceModule.getLatestExternalDatabaseRefID());
      }
    } catch (JSONException e) {
      e.printStackTrace();
      aziFaceModule.sendEvent("onCloseModal", false);
      aziFaceModule.processorPromise.reject("Exception raised while attempting to create JSON payload for upload.",
          "JSONError");
    }

    final String endpoint = this.faceConfig.getEndpoint();
    okhttp3.Request request = new okhttp3.Request.Builder()
        .url(Config.BaseURL + endpoint)
        .headers(Config.getHeaders("POST"))
        .post(new ProgressRequestBody(
            RequestBody.create(MediaType.parse("application/json; charset=utf-8"), parameters.toString()),
            new ProgressRequestBody.Listener() {
              @Override
              public void onUploadProgressChanged(long bytesWritten, long totalBytes) {
                final float uploadProgressPercent = ((float) bytesWritten) / ((float) totalBytes);
                faceScanResultCallback.uploadProgress(uploadProgressPercent);
              }
            }))
        .build();

    NetworkingHelpers.getApiClient().newCall(request).enqueue(new Callback() {
      @Override
      public void onResponse(@NonNull Call call, @NonNull okhttp3.Response response) throws IOException {
        String responseString = response.body().string();
        response.body().close();
        try {
          JSONObject responseJSON = new JSONObject(responseString);
          boolean wasProcessed = responseJSON.getBoolean("wasProcessed");
          String scanResultBlob = responseJSON.getString("scanResultBlob");

          if (wasProcessed) {
            final String message = faceConfig.getSuccessMessage();
            FaceTecCustomization.overrideResultScreenSuccessMessage = aziThemeUtils
                .handleMessage(key, "successMessage", message);
            success = faceScanResultCallback.proceedToNextStep(scanResultBlob);
            if (success) {
              aziFaceModule.sendEvent("onCloseModal", false);
              aziFaceModule.processorPromise.resolve(true);
            }
          } else {
            faceScanResultCallback.cancel();
            aziFaceModule.sendEvent("onCloseModal", false);
            aziFaceModule.processorPromise.reject("AziFace SDK values were not processed!",
                "AziFaceValuesWereNotProcessed");
          }
        } catch (JSONException e) {
          e.printStackTrace();
          faceScanResultCallback.cancel();
          aziFaceModule.sendEvent("onCloseModal", false);
          aziFaceModule.processorPromise.reject("Exception raised while attempting to parse JSON result.",
              "JSONError");
        }
      }

      @Override
      public void onFailure(@NonNull Call call, IOException e) {
        faceScanResultCallback.cancel();
        aziFaceModule.sendEvent("onCloseModal", false);
        aziFaceModule.processorPromise.reject("Exception raised while attempting HTTPS call.", "HTTPSError");
      }
    });
  }

  public boolean isSuccess() {
    return this.success;
  }
}
